xen: Implement support for new kexec load types

The new load types allow kexec to work with Secure Boot.

With the new load types, purgatory is provided by Xen so leave that part out.
For verification purposes, the entire kernel is passed in as the first segment
so rework the loader accordingly.

Signed-off-by: Kevin Lampis <klampis@cloud.com>
Signed-off-by: Gerald Elder-Vass <gerald.elder-vass@cloud.com>
diff --git a/kexec/arch/x86_64/kexec-bzImage64.c b/kexec/arch/x86_64/kexec-bzImage64.c
index aba4e3b8905c..9c03061b4064 100644
--- a/kexec/arch/x86_64/kexec-bzImage64.c
+++ b/kexec/arch/x86_64/kexec-bzImage64.c
@@ -20,6 +20,7 @@
 #define _GNU_SOURCE
 #include <stddef.h>
 #include <stdio.h>
+#include <stdbool.h>
 #include <string.h>
 #include <limits.h>
 #include <stdlib.h>
@@ -105,6 +106,155 @@ void bzImage64_usage(void)
 		);
 }
 
+static int do_xen_bzImage64_load(struct kexec_info *info,
+			const char *kernel, off_t kernel_len,
+			const char *command_line, off_t command_line_len,
+			const char *initrd, off_t initrd_len)
+{
+	struct x86_linux_header setup_header;
+	struct x86_linux_param_header *real_mode;
+	int setup_sects;
+	size_t size;
+	int kern16_size;
+	unsigned long setup_base, setup_size, setup_header_size;
+	char *modified_cmdline;
+	unsigned long align, addr;
+	unsigned kern16_size_needed;
+
+	/*
+	 * Find out about the file I am about to load.
+	 */
+	if (kernel_len < (2 * 512))
+		return -1;
+
+	memcpy(&setup_header, kernel, sizeof(setup_header));
+	setup_sects = setup_header.setup_sects;
+	if (setup_sects == 0)
+		setup_sects = 4;
+	kern16_size = (setup_sects + 1) * 512;
+	if (kernel_len < kern16_size) {
+		fprintf(stderr, "BzImage truncated?\n");
+		return -1;
+	}
+
+	if ((uintmax_t)command_line_len > (uintmax_t)setup_header.cmdline_size) {
+		dbgprintf("Kernel command line too long for kernel!\n");
+		return -1;
+	}
+
+	/* Need to append some command line parameters internally in case of
+	 * taking crash dumps.
+	 */
+	if (info->kexec_flags & (KEXEC_ON_CRASH | KEXEC_PRESERVE_CONTEXT)) {
+		modified_cmdline = xmalloc(COMMAND_LINE_SIZE);
+		memset((void *)modified_cmdline, 0, COMMAND_LINE_SIZE);
+		if (command_line) {
+			strncpy(modified_cmdline, command_line,
+					COMMAND_LINE_SIZE);
+			modified_cmdline[COMMAND_LINE_SIZE - 1] = '\0';
+		}
+
+		/* If panic kernel is being loaded, additional segments need
+		 * to be created. load_crashdump_segments will take care of
+		 * loading the segments as high in memory as possible, hence
+		 * in turn as away as possible from kernel to avoid being
+		 * stomped by the kernel.
+		 */
+		if (load_crashdump_segments(info, modified_cmdline, -1, 0) < 0)
+			return -1;
+
+		/* Use new command line buffer */
+		command_line = modified_cmdline;
+		command_line_len = strlen(command_line) + 1;
+	}
+
+	/* Do not load purgatory. It will be loaded by Xen for security reasons. */
+
+
+	/* The argument/parameter segment */
+	kern16_size_needed = kern16_size;
+	if (kern16_size_needed < 4096)
+		kern16_size_needed = 4096;
+	setup_size = kern16_size_needed + command_line_len + PURGATORY_CMDLINE_SIZE;
+	real_mode = xmalloc(setup_size);
+	memset(real_mode, 0, setup_size);
+
+	/* only copy setup_header */
+	setup_header_size = kernel[0x201] + 0x202 - 0x1f1;
+	if (setup_header_size > 0x7f)
+		setup_header_size = 0x7f;
+	memcpy((unsigned char *)real_mode + 0x1f1, kernel + 0x1f1,
+		 setup_header_size);
+
+	/* No real mode code will be executing. setup segment can be loaded
+	 * anywhere as we will be just reading command line.
+	 */
+	setup_base = add_buffer(info, real_mode, setup_size, setup_size,
+				16, 0x3000, -1, -1);
+
+	dbgprintf("Loaded real_mode_data and command line at 0x%lx\n",
+			setup_base);
+
+
+	/* The main kernel segment */
+	{
+		/* need to use run-time size for buffer searching */
+		size = _ALIGN(real_mode->init_size, 4096);
+		align = real_mode->kernel_alignment;
+
+		/* Traditional kexec chops off the kernel header and only loads
+		 * kernel+kern16_size but for the kernel to cryptographically verify
+		 * integrity we need to load the whole kernel image.
+		 *
+		 * Loading the whole kernel complicates things because kernel+kern16 still
+		 * needs to be aligned with kernel_alignment however this is now in the
+		 * middle of the buffer rather than the start.
+		 *
+		 * Here we find a memory hole bigger than we need and then later we will
+		 * adjust the start address to achieve the correct alignment.
+		 *
+		 *    .- addr
+		 *    v
+		 *    +--------------------------------------+--------------------...
+		 *    | kern16 | kernel+kern16_size          |
+		 *    +--------------------------------------+--------------------...
+		 * 0x100000                               0x200000
+		 */
+		size += real_mode->kernel_alignment;
+		addr = add_buffer(info, kernel, kernel_len, size, align, 0x100000, -1, -1);
+		if (addr == ULONG_MAX)
+			die("can not load bzImage64");
+
+		/*
+		 *                                  .-addr   .-kernel_alignment
+		 *                                  v        v
+		 *    +--------------------------------------+--------------------...
+		 *    | waste                       | kern16 | kernel+kern16_size
+		 *    +--------------------------------------+--------------------...
+		 * 0x100000                               0x200000
+		 */
+		const unsigned long waste = real_mode->kernel_alignment - kern16_size;
+		addr += waste;
+		info->segment[info->nr_segments-1].mem   = (void*)addr;
+		info->segment[info->nr_segments-1].memsz -= waste;
+		info->segment[info->nr_segments-1].is_kernel = true;
+	}
+
+	info->entry = setup_base; // Kernel parameters
+
+
+	/* Tell the kernel what is going on */
+	setup_linux_bootloader_parameters_high(info, real_mode, setup_base,
+			kern16_size_needed, command_line, command_line_len,
+			initrd, initrd_len, 1); /* put initrd high too */
+
+
+	/* Fill in the information BIOS calls would normally provide. */
+	setup_linux_system_parameters(info, real_mode);
+
+	return 0;
+}
+
 static int do_bzImage64_load(struct kexec_info *info,
 			const char *kernel, off_t kernel_len,
 			const char *command_line, off_t command_line_len,
@@ -383,7 +533,11 @@ int bzImage64_load(int argc, char **argv, const char *buf, off_t len,
 	if (ramdisk)
 		ramdisk_buf = slurp_file(ramdisk, &ramdisk_length);
 
-	if (entry_16bit || entry_32bit)
+	if (xen_present() && xen_is_locked_down())
+		result = do_xen_bzImage64_load(info, buf, len, command_line,
+					command_line_len, ramdisk_buf,
+					ramdisk_length);
+	else if (entry_16bit || entry_32bit)
 		result = do_bzImage_load(info, buf, len, command_line,
 					command_line_len, ramdisk_buf,
 					ramdisk_length, 0, 0, entry_16bit);
diff --git a/kexec/kexec-xen.c b/kexec/kexec-xen.c
index a7c89339c4b7..c2b91df9ecf4 100644
--- a/kexec/kexec-xen.c
+++ b/kexec/kexec-xen.c
@@ -62,6 +62,7 @@ int __xc_interface_close(xc_interface *xch)
 
 	return rc;
 }
+
 #endif /* CONFIG_LIBXENCTRL_DL */
 
 int xen_get_kexec_range(int range, uint64_t *start, uint64_t *end)
@@ -91,14 +92,47 @@ out:
 	return rc;
 }
 
+bool xen_is_locked_down(void)
+{
+	static int locked_down = -1;
+	FILE *f;
+	char buf[6];
+
+	if (locked_down >= 0)
+		goto out;
+
+	f = fopen("/sys/kernel/security/lockdown", "r");
+	if (!f) {
+		locked_down = 0;
+		goto out;
+	}
+	if (fread(buf, 1, sizeof(buf), f) != sizeof(buf)) {
+		fclose(f);
+		locked_down = 0;
+		goto out;
+	}
+	fclose(f);
+	locked_down = strcmp(buf, "[none]") ? 1 : 0;
+
+out:
+	return locked_down == 1;
+}
+
 static uint8_t xen_get_kexec_type(unsigned long kexec_flags)
 {
-	if (kexec_flags & KEXEC_ON_CRASH)
-		return KEXEC_TYPE_CRASH;
+	if (kexec_flags & KEXEC_ON_CRASH) {
+		if (xen_is_locked_down())
+			return KEXEC_TYPE_CRASH_EFI;
+		else
+			return KEXEC_TYPE_CRASH;
+	}
 
 	if (kexec_flags & KEXEC_LIVE_UPDATE)
 		return KEXEC_TYPE_LIVE_UPDATE;
 
+	if (xen_is_locked_down())
+		return KEXEC_TYPE_CRASH_EFI;
+
 	return KEXEC_TYPE_DEFAULT;
 }
 
@@ -113,7 +147,8 @@ int xen_kexec_load(struct kexec_info *info)
 	xc_hypercall_buffer_array_t *array = NULL;
 	uint8_t type;
 	uint8_t arch;
-	xen_kexec_segment_t *xen_segs, *seg;
+	xen_kexec_segment_t *xen_segs, *seg, swap_seg;
+	xen_kexec_segment_t *kernel_seg = NULL;
 	int s;
 	int ret = -1;
 
@@ -175,6 +210,10 @@ int xen_kexec_load(struct kexec_info *info)
 		seg->buf_size = segments[s].bufsz;
 		seg->dest_maddr = (uint64_t)segments[s].mem;
 		seg->dest_size = segments[s].memsz;
+
+		if (segments[s].is_kernel)
+			kernel_seg = seg;
+
 		seg++;
 
 		low_watermark = (uint64_t)segments[s].mem + segments[s].memsz;
@@ -188,6 +227,12 @@ int xen_kexec_load(struct kexec_info *info)
 		seg++;
 	}
 
+	if (kernel_seg) {
+		swap_seg = *kernel_seg;
+		*kernel_seg = *xen_segs;
+		*xen_segs = swap_seg;
+	}
+
 	type = xen_get_kexec_type(info->kexec_flags);
 
 	arch = (info->kexec_flags & KEXEC_ARCH_MASK) >> 16;
@@ -196,8 +241,8 @@ int xen_kexec_load(struct kexec_info *info)
 		arch = EM_386;
 #endif
 
-	ret = xc_kexec_load(xch, type, arch, (uint64_t)info->entry,
-			    nr_segments + nr_low_segments, xen_segs);
+        ret = xc_kexec_load(xch, type, arch, (uint64_t)info->entry,
+                              nr_segments + nr_low_segments, xen_segs);
 
 out:
 	xc_hypercall_buffer_array_destroy(xch, array);
diff --git a/kexec/kexec-xen.h b/kexec/kexec-xen.h
index 70fb576e93f3..7edde2de7683 100644
--- a/kexec/kexec-xen.h
+++ b/kexec/kexec-xen.h
@@ -78,6 +78,14 @@ extern int __xc_interface_close(xc_interface *xch);
 #define KEXEC_TYPE_LIVE_UPDATE 2
 #endif
 
+#ifndef KEXEC_TYPE_DEFAULT_EFI
+#define KEXEC_TYPE_DEFAULT_EFI 2
+#endif
+
+#ifndef KEXEC_TYPE_CRASH_EFI
+#define KEXEC_TYPE_CRASH_EFI   3
+#endif
+
 #ifndef KEXEC_LIVE_UPDATE
 #define KEXEC_LIVE_UPDATE 0x00000004
 #endif
diff --git a/kexec/kexec.c b/kexec/kexec.c
index 7c614b08b6c5..19e2587d884c 100644
--- a/kexec/kexec.c
+++ b/kexec/kexec.c
@@ -25,6 +25,7 @@
 #include <stdarg.h>
 #include <string.h>
 #include <stdlib.h>
+#include <stdbool.h>
 #include <errno.h>
 #include <limits.h>
 #include <sys/ioctl.h>
@@ -365,6 +366,7 @@ void add_segment_phys_virt(struct kexec_info *info,
 	info->segment[info->nr_segments].bufsz = bufsz;
 	info->segment[info->nr_segments].mem   = (void *)base;
 	info->segment[info->nr_segments].memsz = memsz;
+	info->segment[info->nr_segments].is_kernel = false;
 	info->nr_segments++;
 	if (info->nr_segments > KEXEC_MAX_SEGMENTS) {
 		fprintf(stderr, "Warning: kernel segment limit reached. "
diff --git a/kexec/kexec.h b/kexec/kexec.h
index 31c323f6746d..f1c6db424da2 100644
--- a/kexec/kexec.h
+++ b/kexec/kexec.h
@@ -6,6 +6,7 @@
 #include <sys/types.h>
 #include <stdio.h>
 #include <stdint.h>
+#include <stdbool.h>
 #include <string.h>
 #define USE_BSD
 #include <byteswap.h>
@@ -127,6 +128,7 @@ struct kexec_segment {
 	size_t bufsz;
 	const void *mem;
 	size_t memsz;
+	bool is_kernel;
 };
 
 struct memory_range {
@@ -337,6 +339,7 @@ int xen_kexec_load(struct kexec_info *info);
 int xen_kexec_unload(uint64_t kexec_flags);
 int xen_kexec_exec(uint64_t kexec_flags);
 int xen_kexec_status(uint64_t kexec_flags);
+bool xen_is_locked_down(void);
 
 extern unsigned long long get_kernel_sym(const char *text);
 
